test/addons-napi/test_promise/test_promise.c

```diff
@@ -0,0 +1,65 @@
+#include <node_api.h>
+#include "../common.h"
+
+napi_ref deferred_ref = NULL;
+
+napi_value createPromise(napi_env env, napi_callback_info info) {
+  napi_value promise, deferred, in_one_call;
+  bool one_call;
+  size_t argc = 1;
+
+  NAPI_CALL(env, napi_get_cb_info(env, info, &argc, &in_one_call, NULL, NULL));
+  NAPI_CALL(env, napi_get_value_bool(env, in_one_call, &one_call));
+
+  if (deferred_ref != NULL) {
+    NAPI_CALL(env, napi_delete_reference(env, deferred_ref));
+  }
+
+  if (one_call) {
+    NAPI_CALL(env, napi_create_promise(env, &deferred, &promise));
+  } else {
+    NAPI_CALL(env, napi_create_promise(env, &deferred, NULL));
+    NAPI_CALL(env, napi_get_promise_from_deferred(env, deferred, &promise));
+  }
+  NAPI_CALL(env, napi_create_reference(env, deferred, 1, &deferred_ref));
+
+  return promise;
+}
```

Comment:
From https://github.com/nodejs/node/pull/13717#issuecomment-309044219, it seems like V8 treats a Resolver just like a Promise, which means that if the user passes the deferred (backed by a Resolver) back into JS the separation between deferred and promise would not be enforced, and the problem you outlined in https://github.com/nodejs/node/pull/13717#issuecomment-309136752 would still be there